File: python-sdk/src/forta_agent/label.py
from enum import IntEnum
from .utils import assert_enum_value_in_dict, assert_non_empty_string_in_dict


class EntityType(IntEnum):
    Unknown = 0
    Address = 1
    Transaction = 2
    Block = 3
    Url = 4


class Label:
    def __init__(self, dict):
        entityTypeVal = dict.get('entity_type', dict.get('entityType'))
        self.entity_type = EntityType[entityTypeVal.title()] if type(
            entityTypeVal) == str else EntityType(entityTypeVal)
        assert_enum_value_in_dict(self.__dict__, 'entity_type', EntityType)
        assert_non_empty_string_in_dict(dict, 'entity')
        self.entity = dict['entity']
        self.confidence = dict['confidence']
        self.label = dict['label']
        self.remove = dict.get('remove', False)
        self.unique_key = dict.get('unique_key', dict.get('uniqueKey'))
        self.metadata = dict.get('metadata') if dict.get(
            'metadata') is not None else {}
        # if metadata is array, convert to map
        if type(self.metadata) is list:
            self.metadata_array_to_map()
        self.id = dict.get('id')
        self.source = LabelSource(dict.get('source')) if dict.get(
            'source') is not None else None
        self.created_at = dict.get('createdAt', dict.get('created_at'))
        self.embedding = dict.get('embedding')

    def toDict(self):
        d = dict(self.__dict__, **{
            'entityType': self.entity_type,
            'uniqueKey': self.unique_key,
        })
        return {k: v for k, v in d.items() if v is not None}

    def metadata_array_to_map(self):
        # convert string array to string key/value map using first '=' character as separator
        # (label metadata is received as string array from API)
        metadata_map = {}
        for item in self.metadata:
            separator_index = item.find('=')
            key = item[0:separator_index]
            value = item[separator_index+1:len(item)]
            metadata_map[key] = value
        self.metadata = metadata_map


class LabelSource:
    def __init__(self, dict):
        self.alert_hash = dict.get('alertHash')
        self.alert_id = dict.get('alertId')
        self.bot = LabelSourceBot(dict.get('bot')) if dict.get(
            'bot') is not None else None
        self.chain_id = dict.get('chainId')
        self.id = dict.get('id')


class LabelSourceBot:
    def __init__(self, dict):
        self.id = dict.get('id')
        self.image = dict.get('image')
        self.image_hash = dict.get('imageHash')
        self.manifest = dict.get('manifest')

File: sdk/label.ts
export enum EntityType {
  Unknown,
  Address,
  Transaction,
  Block,
  Url,
}

export const ENTITY_TYPE_STRING_TO_ENUM = {
  UNKNOWN: EntityType.Unknown,
  ADDRESS: EntityType.Address,
  TRANSACTION: EntityType.Transaction,
  BLOCK: EntityType.Block,
  URL: EntityType.Url,
};

export type LabelSource = {
  alertHash?: string;
  alertId?: string;
  bot?: {
    id: string;
    image: string;
    imageHash: string;
    manifest: string;
  };
  chainId?: number;
  id?: string;
};

type LabelInput = {
  entityType: EntityType;
  entity: string;
  label: string;
  confidence: number;
  remove?: boolean;
  metadata?: { [key: string]: string };
  source?: LabelSource;
  createdAt?: string;
  id?: string;
  uniqueKey?: string;
  embedding?: number[];
};

export class Label {
  private constructor(
    readonly entityType: EntityType,
    readonly entity: string,
    readonly label: string,
    readonly confidence: number,
    readonly remove: boolean,
    readonly metadata: { [key: string]: string },
    readonly source?: LabelSource,
    readonly id?: string,
    readonly createdAt?: string,
    readonly uniqueKey?: string,
    readonly embedding?: number[]
  ) {}

  static from(labelInput: LabelInput) {
    return this.fromObject(labelInput);
  }

  static fromObject({
    entityType,
    entity,
    label,
    confidence,
    remove = false,
    metadata = {},
    source,
    id,
    createdAt,
    uniqueKey,
    embedding
  }: LabelInput) {
    if (typeof entityType == "string") {
      entityType = ENTITY_TYPE_STRING_TO_ENUM[entityType];
    }
    if (Array.isArray(metadata)) {
      // convert string array to string key/value map using first '=' character as separator
      // (label metadata is received as string array for handleAlert)
      let metadataMap: { [key: string]: string } = {};
      for (const arrayItem of metadata) {
        const separatorIndex = arrayItem.indexOf("=");
        const key = arrayItem.substring(0, separatorIndex);
        const value = arrayItem.substring(separatorIndex + 1, arrayItem.length);
        metadataMap[key] = value;
      }
      metadata = metadataMap;
    }
    return new Label(
      entityType,
      entity,
      label,
      confidence,
      remove,
      metadata,
      source,
      id,
      createdAt,
      uniqueKey,
      embedding
    );
  }
}
